### PR TITLE
Add sats as a native currency

### PR DESCRIPTION
**Pull request by Kukks**
This will allow you to create an invoice where its primary currency is denominated in sats

### CODE REVIEW

**Comment by Kukks** [comment]
@opennoms  This is what you wanted right?

**Comment by Kukks** [comment]
oops @openoms *

**Comment by NicolasDorier** [comment]
this is not good. If an exchange can solve USD_BTC, then it should bt able to solve USD_SATS. Basically it should just be an implicit rate rule.

**Comment by Kukks** [comment]
cleaned iy up to only use the rates scripting engine

File: BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs [hunk 1]
```diff
@@ -22,6 +22,11 @@ public void InitBitcoin()
                 UriScheme = "bitcoin",
                 CryptoImagePath = "imlegacy/bitcoin.svg",
                 LightningImagePath = "imlegacy/bitcoin-lightning.svg",
+                DefaultRateRules = new[]
+                {
+                    "SATS_X = SATS_BTC * BTC_X",
```
Comment (NicolasDorier) [inline BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs]: you should not do this, look at the implicit rules like `1/X`, you should modify the  the RateRules class.

Comment (Kukks) [reply BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs]: RateRules class is so complex, where are these implicit rules

**Comment by bolatovumar** [comment]
I kinda wanted this as well and thought of working on it but good to know someone is working on it.

**Comment by dennisreimann** [comment]
This will be great for donations pages. Had to fiddle around to make mine handle BTC fractions.

File: BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs (hunk 1 above)
Comment (Kukks) [reply BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs]: ping @NicolasDorier

**Comment by NicolasDorier** [comment]
Can't be a default rules, as default rules can be overidden.

I can't think of a way to implement this which would not result in crazy corner case.

**Comment by Kukks** [comment]
> 
> 
> Can't be a default rules, as default rules can be overidden.
> 
> I can't think of a way to implement this which would not result in crazy corner case.

that's fine though, if someone wants to override the rule then it's a deliberate action?

**Comment by NicolasDorier** [comment]
Maybe by changing `FindBestCandidate` such that, if the pair is `SATS_X` or `X_STATS`, then the rule is replaced by `BTC_X` or `X_BTC`.
Then later we take the matched expression * 10^8

**Comment by NicolasDorier** [comment]
> that's fine though, if someone wants to override the rule then it's a deliberate action?

No. This will create crazy corner case that will take ages to debug and support. And most likely the person will do it by mistake when he wants to create his own script.

**Comment by Kukks** [comment]
How about this now?

File: BTCPayServer.Rating/RateRules.cs [hunk 2]
```diff
@@ -100,6 +100,8 @@ public SyntaxNode GetSyntaxNode()
         RuleList ruleList;
 
         decimal _Spread;
+        private const string ImplicitSatsRule = "SATS_X = SATS_BTC * BTC_X;\nSATS_BTC = 0.00000001;\n";
```
Comment (NicolasDorier) [inline BTCPayServer.Rating/RateRules.cs]: Only use `SATS_BTC = 0.00000001`, inverses are implicit.

Also you need to test that properly

File: BTCPayServer.Rating/RateRules.cs [hunk 3]
```diff
@@ -127,6 +129,7 @@ public static bool TryParse(string str, out RateRules rules)
         }
         public static bool TryParse(string str, out RateRules rules, out List<RateRulesErrors> errors)
         {
+            str = ImplicitSatsRule + str;
```
Comment (NicolasDorier) [inline BTCPayServer.Rating/RateRules.cs]: Mofidy the rule list instead.

Comment (Kukks) [reply BTCPayServer.Rating/RateRules.cs]: The Rule list is black magic to me, no clue what or how it does anything

File: BTCPayServer.Rating/RateRules.cs (hunk 2 above)
Comment (Kukks) [reply BTCPayServer.Rating/RateRules.cs]: Are you sure? As soon as I remove it, it makes every other rate for it unavailable.

**Comment by NicolasDorier** [comment]
Write tests about it, then I can try to tweak it the correct way

**Comment by Kukks** [comment]
rebased and added a test

**Comment by NicolasDorier** [comment]
still need test

**Comment by pavlenex** [comment]
tACK, though I must admit I first tried  to create an invoice with `SAT` which made more sense, but then realize `SATS` is needed.

**Pull request merged**